spec/integration_spec.rb

```diff
@@ -126,4 +126,43 @@
       end
     end
   end
+
+  context "performance" do
+    let(:user) { UserWithOneServerAndDeepPath.new }
+    let(:padded_user_id) { user.id.to_s.rjust(3, "0") }
+    let(:uploaded_file_deep_path) { FtpServer::USER1_PATH + "/img/user_with_one_server_and_deep_paths/avatars/000/000/#{padded_user_id}/original/avatar.jpg" }
+
+    it "triggers minimal amount of ftp commands by using #put_files internally" do
+      expect_any_instance_of(Net::FTP).to receive(:nlst).exactly(7).times.and_call_original
+      expect_any_instance_of(Net::FTP).to receive(:mkdir).exactly(9).times.and_call_original
+      user.avatar = file
+      user.save!
+      File.exist?(uploaded_file_deep_path).should be true
+    end
+
+    context "legacy" do
+      it "triggers plenty of MKDIR commands when a) not using #put_files and b) existence is not checked beforehand" do
+        expect_any_instance_of(Hash).to receive(:length).exactly(4).times.and_return(1) # ugly hack to enforce legacy behaviour
+        expect_any_instance_of(Paperclip::Storage::Ftp::Server).to receive(:file_exists?).exactly(3).times.and_return(false)
+
+        expect_any_instance_of(Net::FTP).to_not receive(:mktree)
+        expect_any_instance_of(Net::FTP).to_not receive(:nlst)
+        expect_any_instance_of(Net::FTP).to receive(:mkdir).exactly(24).times.and_call_original
+
+        user.avatar = file
+        user.save!
+      end
+
+      it "triggers plenty of ftp commands when not using #put_files" do
+        expect_any_instance_of(Hash).to receive(:length).exactly(4).times.and_return(1) # ugly hack to enforce legacy behaviour
+
+        expect_any_instance_of(Net::FTP).to_not receive(:mktree)
+        expect_any_instance_of(Net::FTP).to receive(:nlst).exactly(3).times.and_call_original
+        expect_any_instance_of(Net::FTP).to receive(:mkdir).exactly(24).times.and_call_original
+
+        user.avatar = file
+        user.save!
+      end
+    end
+  end
```

Comment:
I appreciate that you took the effort to create these specs, but the integration spec should really only verify the general functionality of the gem, not internal details. Also we don't need to keep legacy behaviour if the new implementation can deal with all cases.